streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java

```diff
@@ -0,0 +1,428 @@
+            ValueAndTimestamp<Agg> rightWinAgg = null;
+
+            // flag to help determine if out-of-order record’s right window is non-empty
+            boolean foundLeftFirst = false;
+            //if current record's left/right windows already exist
+            boolean leftWinExists = false;
+            boolean rightWindowExists = false;
+            //to determine if we're creating the previous record's right window, helps with determining empty windows
+            boolean prevRightWinExists = false;
+            boolean prevRightWinAlreadyCreated = false;
+
+            try (
+                    //Fetch all the windows that have a start time between timestamp and timestamp+timeDifference
+                    final KeyValueIterator<Windowed<K>, ValueAndTimestamp<Agg>> iterator = windowStore.fetch(key,
+                            key,
+                            Instant.ofEpochMilli(timestamp - 2 * windows.timeDifference()),
+                            Instant.ofEpochMilli(timestamp + 1))
+            ) {
+                KeyValue<Windowed<K>, ValueAndTimestamp<Agg>> next;
+                //if we've already seen the window with the closest start time to the record
+                boolean foundFirst = false;
+                //if we've already seen the window with the closest end time to the record
+                boolean foundFirstEndTime = false;
+                while (iterator.hasNext()) {
+                    next = iterator.next();
+                    windowStartTimes.add(next.key.window().start());
+
+                    //determine if current record's right window exists
+                    if (next.key.window().start() == timestamp + 1) {
+                        rightWindowExists = true;
+                        continue;
+                    } else if (next.key.window().end() > timestamp) {
+                        if (!foundFirst) {
+                            foundFirst = true;
+                            if (isLeftWindow(next)) {
+                                foundLeftFirst = true;
+                            }
+                        }
+                        rightWinAgg = next.value;
+                        putAndForward(next.key.window(), next.value, key, value, closeTime, timestamp);
+                        continue;
+                    } else if (next.key.window().end() == timestamp) {
+                        putAndForward(next.key.window(), next.value, key, value, closeTime, timestamp);
+                        leftWinExists = true;
+                        continue;
+                    } else {
+                        if (!foundFirstEndTime) {
+                            leftWinAgg = next.value;
+                            foundFirstEndTime = true;
+                        }
+                        if (prevRightWinExists) {
+                            break;
+                        }
+                        if (isLeftWindow(next)) {
```

Comment:
Can we add a comment to clarify that we're checking whether it's a left window because that tells us there was a record at this window's end time